pkg/controllers/state/statenodepool.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+)
+
+// Currently tracks in-flight NodeClaims for a NodePool
+type NodePooltracker struct {
+	ReservedNodeLimit atomic.Int64
+}
+
+type NodeClaimState struct {
+	Active   sets.Set[string]
+	Deleting sets.Set[string]
+}
+
+// StateNodePool is a cached version of a NodePool in the cluster that maintains state which is expensive to compute every time it's needed.
+type NodePoolState struct {
+	mu sync.RWMutex
+
+	nodePoolNameToNodeClaimState map[string]NodeClaimState   // node pool name -> node claim state (Running and Deleting node claim names)
+	nodeClaimNameToNodePoolName  map[string]string           // node claim name -> node pool name
+	nodePoolNameToNodePoolLimit  map[string]*NodePooltracker // node pool -> nodepool limit
```

Comment:
```suggestion
	nodePoolNameToReservedNodeCount  map[string]atomic.Int64// node pool -> nodepool limit
```

IMO this is what we should use for the type and the naming